File: agentkit/ts/src/tools/payment.ts
/**
 * requestPayment: a thin agentkit -> agentpay bridge.
 *
 * Posts to `${baseUrl}/v1/sessions` and returns the JSON payload. Pass a
 * custom fetchImpl in tests; in production it uses the global fetch.
 */
import type { ToolSpec, ToolHandler, ToolContext } from "./registry";

export interface RequestPaymentOptions {
  baseUrl: string;
  apiKey?: string;
  fetchImpl?: typeof fetch;
}

export interface RequestPaymentTool {
  spec: ToolSpec;
  handler: ToolHandler;
}

export function requestPayment(opts: RequestPaymentOptions): RequestPaymentTool {
  const spec: ToolSpec = {
    name: "request_payment",
    description: "Request a payment via agentpay. Returns the checkout URL.",
    parameters: {
      type: "object",
      properties: {
        amount: { type: "integer", description: "Amount in the smallest currency unit (cents)." },
        description: { type: "string", description: "Short reason shown to the payer." },
      },
      required: ["amount", "description"],
    },
    risk: "write",
  };

  const fetchImpl = opts.fetchImpl ?? globalThis.fetch;

  const handler: ToolHandler = async (_ctx: ToolContext, args: Record<string, unknown>) => {
    const headers: Record<string, string> = { "Content-Type": "application/json" };
    if (opts.apiKey) headers.Authorization = `Bearer ${opts.apiKey}`;
    const url = `${opts.baseUrl.replace(/\/$/, "")}/v1/sessions`;
    const r = await fetchImpl(url, { method: "POST", headers, body: JSON.stringify(args) });
    if (!r.ok) {
      throw new Error(`agentpay POST /v1/sessions failed: ${r.status} ${r.statusText}`);
    }
    return (await r.json()) as Record<string, unknown>;
  };

  return { spec, handler };
}

File: agentkit/sdk/agentkit/runtime/tools/payment.py
"""`request_payment`: the agentpay bridge tool.

Returns a (ToolSpec, handler) pair. The handler POSTs the args to
``<base_url>/v1/sessions`` and returns the parsed JSON. Pass a custom httpx
client in tests; in production a fresh AsyncClient is created per call.
"""
from __future__ import annotations

from typing import Any, Awaitable, Callable

import httpx

from agentkit.runtime.tools.registry import ToolSpec


PaymentHandler = Callable[[Any, dict[str, Any]], Awaitable[Any]]


def RequestPaymentTool(
    *,
    base_url: str,
    api_key: str | None = None,
    client: httpx.AsyncClient | None = None,
) -> tuple[ToolSpec, PaymentHandler]:
    """Build the agentpay bridge tool.

    Parameters
    ----------
    base_url:
        Root URL of the agentpay API (no trailing slash required).
    api_key:
        Optional bearer token. When provided, sent as ``Authorization`` header.
    client:
        Optional pre-built httpx AsyncClient. Useful in tests with MockTransport.
        When omitted, the handler creates a one-shot AsyncClient per call.
    """
    base = base_url.rstrip("/")

    spec = ToolSpec(
        name="request_payment",
        description="Request a payment via agentpay. Returns the checkout URL.",
        parameters={
            "type": "object",
            "properties": {
                "amount": {
                    "type": "integer",
                    "description": "Amount in the smallest currency unit (cents).",
                },
                "description": {
                    "type": "string",
                    "description": "Short reason shown to the payer.",
                },
            },
            "required": ["amount", "description"],
        },
        risk="write",
    )

    headers = {"Content-Type": "application/json"}
    if api_key:
        headers["Authorization"] = f"Bearer {api_key}"

    async def handler(ctx: Any, args: dict[str, Any]) -> Any:
        url = f"{base}/v1/sessions"

        async def _post(c: httpx.AsyncClient) -> Any:
            r = await c.post(url, headers=headers, json=args)
            if r.status_code >= 400:
                raise RuntimeError(
                    f"agentpay POST /v1/sessions failed: {r.status_code} {r.reason_phrase}"
                )
            return r.json()

        if client is not None:
            return await _post(client)
        async with httpx.AsyncClient() as c:
            return await _post(c)

    return spec, handler
